docs/WASI-design-principles.md

```diff
@@ -0,0 +1,93 @@
+
+All this said, maximal POSIX conformance is not WASI's primary goal. Some reasons include:
+
+ - `fork` -- It's not that we can't make `fork` work -- we can, it's that `fork` carries with it the assumption of copy-on-write memory optimizations which won't be feasible in many environments where we want to run WASI, such as nano-processes. There may eventually be compatibility layers that provide `fork` to help people port POSIX code to WASI, however there's a difference between providing `fork` as an optional compatibility layer and having it as a cornerstone of an ecosystem as it is in POSIX.
+
+   And when we take `fork` out of the focus, it changes the way we think about a lot of other things, such as `execve`, `dup`, `fcntl`-style file locking, and even processes.
+
+ - Users and Groups -- POSIX's Users and Groups subsystem are notoriously inflexible, to the point where much of the computing world has moved to using containers and VMs and other forms of single-user environments because traditional multi-user OS functionality doesn't do what's needed.
+
+   And, when running untrusted code, it isn't desirable to run it as a user's normal identity, because it shouldn't inherit all of the rights a user has, but it also doesn't help to run it as user "nobody", as it's still useful to grant it some rights and restrict it from others.
+
+   And, we are aiming for portability to OS's which don't have POSIX-style users and groups, and systems which don't have OS's at all.
+
+ - Asynchronous signals and handlers -- The core WebAssembly semantics don't support these, which would need to change before WASI could consider supporting them, and there are currently no proposals for doing so. In POSIX, some interfaces are designed with the assumption that signals like `SIGPIPE`, `SIGALRM`, `SIGCHLD`, `SIGIO` and others exist and can cover certain situations, so in the absence of these signals, those interfaces won't always make sense.
+
+ - Shared filesystem views - One of the unique capabilities WebAssembly brings to the table is the possibility of shared-nothing linking between applications and libraries. Shared-nothing means that all communication is via explicit calls, and the libraries don't share an address space or any other implicit shared state. But if we run both sides within the same filesystem view, that would give them a large body of shared state. Union mounts, mandatory access control systems, user namespaces, and other techniques can help, but often require complex configuration, heavy-weight boundaries, and sometimes even admin privileges to set up.
+
+This has wide-ranging implications. Much of POSIX is oriented around passing around strings, whether through command-line arguments, environment variables, or paths embedded in files, with the assumption that there's a shared filesystem view between components. As we said above, we're de-emphasizing strings, which dovetails with de-emphasizing shared filesystem views.
```

Comment:
It might be useful to explain what pattern WASI will use to share files and data in the absence of a shared filesystem view. Will applications be expected to explicitly pass file access capabilities for each relevant file to each relevant dependency?

Reply:
Good idea. I've now added some extra text to illustrate this.